packages/grafana-data/src/types/live.ts

```diff
@@ -0,0 +1,172 @@
+   */
+  variables?: Array<SelectableValue<string>>;
+
+  /**
+   * The channel keeps track of who else is connected to the same channel
+   */
+  hasPresense?: boolean;
+
+  /**
+   * This method will be defined if it is possible to publish in this channel.
+   * The function will return true/false if the current user can publish
+   */
+  canPublish?: () => boolean;
+
+  /** convert the raw stream message into a message that should be broadcast */
+  processMessage?: (msg: any) => TMessage;
+}
+
+export enum LiveChannelConnectionState {
+  /** The connection is not yet established */
+  Pending = 'pending',
+  /** Connected to the channel */
+  Connected = 'connected',
+  /** Disconnected from the channel.  The channel will reconnect when possible */
+  Disconnected = 'disconnected',
+  /** Was at some point connected, and will not try to reconnect */
+  Shutdown = 'shutdown',
+  /** Channel configuraiton was invalid and will not connect */
+  Invalid = 'invalid',
+}
+
+/**
+ * @experimental
+ */
+export interface LiveChannelStatus {
+  /**
+   * {scope}/{namespace}/{path}
+   */
+  id: string;
+
+  /**
+   * unix millies timestamp for the last status change
+   */
+  timestamp: number;
+
+  /**
+   * flag if the channel is activly connected to the channel.
+   * This may be false while the connections get established or if the network is lost
+   * As long as the `shutdown` flag is not set, the connection will try to reestablish
+   */
+  state: LiveChannelConnectionState;
+
+  /**
+   * The last error.
+   *
+   * This will remain in the status until a new message is succesfully recieved from the channel
+   */
+  error?: any;
+}
+
+/**
+ * @experimental
+ */
+export interface LiveChannelJoinLeave {
+  user: any;
+}
+
+/**
+ * @experimental
+ */
+export interface LiveChannelPresense {
+  users: any;
+}
+
+export interface LiveChannelMessage<TMessage = any> {
+  type: 'status' | 'message' | 'join' | 'leave';
+  message: TMessage | LiveChannelStatus | LiveChannelJoinLeave;
+}
+
+/**
+ * @experimental
+ */
+export interface LiveChannel<TMessage = any, TPublish = any> {
+  /** The fully qualified channel id: ${scope}/${namespace}/${path} */
+  id: string;
+
+  /** The scope for this channel */
+  scope: LiveChannelScope;
+
+  /** datasourceId/plugin name/feature depending on scope */
+  namespace: string;
+
+  /** additional qualifier */
+  path: string;
+
+  /** Unix timestamp for when the channel connected */
+  opened: number;
+
+  /** Static definition of the channel definition.  This may describe the channel usage */
+  config?: LiveChannelConfig;
+
+  /**
+   * Watch all events in this channel
+   */
+  getStream: () => Observable<LiveChannelMessage<TMessage>>;
+
+  /**
+   * For channels that support presense, this will request the current state from the server.
+   *
+   * Join and leave messages will be sent to the open stream
+   */
+  getPresense?: () => Promise<LiveChannelPresense>;
```

Comment:
getStream & getPresense should be merged I think. Or is there any good reason that I've to subscribe to both?

Reply:
maybe -- in the centrifuge API, getPresense is an async call that gets the current state, and then the join/leave messages are deltas.  

We could potentially wrap that all and return a full managed list of users on join/leave... but again, I am not totally sure what pattern makes the most sense until we try it :)